jni/src/common/Foreign/JNI/Unsafe.hs

```diff
@@ -677,12 +712,26 @@ getMethodID cls methodname (coerce -> sig) = throwIfJNull cls $
                                      $(char *methodnamep),
                                      $(char *sigp)) } |]
 
+-- | wrapper around getStaticMethodID' : raise a verbose exception if no method was found
 getStaticMethodID
   :: JClass -- ^ A class object as returned by 'findClass'
   -> JNI.String -- ^ Field name
   -> MethodSignature -- ^ JNI signature
   -> IO JMethodID
-getStaticMethodID cls methodname (coerce -> sig) = throwIfJNull cls $
+getStaticMethodID cls methodName sig = catch
+  (getStaticMethodID' cls methodName sig)
+  (\(e::JVMException) -> do
+    signatures <- getSignatures cls methodName
+    className <- getClassName cls
+    throwIO $ VerboseNoSuchMethodException className methodName signatures
+  )
+
+getStaticMethodID'
+  :: JClass -- ^ A class object as returned by 'findClass'
+  -> JNI.String -- ^ Field name
+  -> MethodSignature -- ^ JNI signature
+  -> IO JMethodID
+getStaticMethodID' cls methodname (coerce -> sig) = throwIfJNull cls $
```

Comment:
How about moving everything in ``Foreign.JNI.Unsafe` to `Foreign.JNI.Unsafe.Internal`, and have `Foreign.JNI.Unsafe` reexport it?

In that way we can separate the JNI bindings from the error handling code that we are adding here.